core/lib/zksync_core/src/fee_model.rs

```diff
@@ -0,0 +1,143 @@
+use std::{fmt, sync::Arc};
+
+use zksync_types::{
+    fee_model::{
+        BatchFeeInput, L1PeggedBatchFeeModelInput, MainNodeFeeModelConfig,
+        MainNodeFeeModelConfigV2, MainNodeFeeParams, MainNodeFeeParamsV1, MainNodeFeeParamsV2,
+        PubdataIndependentBatchFeeModelInput,
+    },
+    U256,
+};
+
+use crate::l1_gas_price::L1GasPriceProvider;
+
+/// Trait responsible for providing fee info for a batch
+pub trait BatchFeeModelInputProvider: fmt::Debug + 'static + Send + Sync {
+    /// Returns the batch fee with scaling applied. This may be used to account for the fact that the L1 gas and pubdata prices may fluctuate, esp.
+    /// in API methods that should return values that are valid for some period of time after the estimation was done.
+    fn get_batch_fee_input_scaled(
+        &self,
+        l1_gas_price_scale_factor: f64,
+        l1_pubdata_price_scale_factor: f64,
+    ) -> BatchFeeInput {
+        let params = self.get_fee_model_params();
+
+        match params {
+            MainNodeFeeParams::V1(params) => {
+                compute_batch_fee_model_input_v1(params, l1_gas_price_scale_factor)
+            }
+            MainNodeFeeParams::V2(params) => compute_batch_fee_model_input_v2(
+                params,
+                l1_gas_price_scale_factor,
+                l1_pubdata_price_scale_factor,
+            ),
+        }
+    }
+
+    /// Returns the batch fee input as-is, i.e. without any scaling for the L1 gas and pubdata prices.
+    fn get_batch_fee_input(&self) -> BatchFeeInput {
+        self.get_batch_fee_input_scaled(1.0, 1.0)
+    }
+
+    /// Returns the fee model parameters.
+    fn get_fee_model_params(&self) -> MainNodeFeeParams;
+}
+
+/// The struct that represents the batch fee input provider to be used in the main node of the server, i.e.
+/// it explicitly gets the L1 gas price from the provider and uses it to calculate the batch fee input instead of getting
+/// it from other node.
+#[derive(Debug)]
+pub(crate) struct MainNodeFeeInputProvider<G: ?Sized> {
+    provider: Arc<G>,
+    config: MainNodeFeeModelConfig,
+}
```

Comment:
It probably makes sense to use `dyn L1GasPriceProvider` instead of generics like we do in other places, cc @popzxc

Reply:
makes sense

Reply:
Yes please.